File: resources/MTAW/client/data/gui/js/crosshair.js
"use strict";
var Crosshair = {
	element : null,
	visible : true,
	
	init : function() {
		Crosshair.element = $( '#crosshair' );
	},
	
	// Показать курсор (вызывается при старте клиенте, чтобы во время загрузки не было видно его)
	enable : function() {
		$( '#crosshair' ).addClass( 'enabled' );
	},
	
	setVisible : function( visible ) {
		if ( visible ) {
			Crosshair.element.removeClass( 'hidden' );
			Crosshair.visible = true;
		} else {
			Crosshair.element.addClass( 'hidden' );
			Crosshair.visible = false;
		}
	},
	
	// Показать подсказку возле курсора. Если showActionKey установлен в true, будет показана кнопка взаимодействия
	setLabel : function( labelText, labelDescription, labelColor, showActionKey, actionKey ) {
		if ( showActionKey == true ) {
			$( '#crosshair-label' ).addClass( 'action-key-visible' );
		} else {
			$( '#crosshair-label' ).removeClass( 'action-key-visible' );
		}
		
		//console.log( labelText, labelDescription, labelColor, showActionKey, actionKey );
		
		if ( actionKey == false || actionKey == undefined || actionKey == null ) {
			$( '#crosshair-label-action-key' ).text( 'E' );
		} else {
			if ( actionKey == "RMB" ) {
				// Правая кнопка мыши
				$( '#crosshair-label-action-key' ).html( "<div class='action-key-rmb' />" );
			} else if ( actionKey == "LMB" ) {
				// Левая кнопка мыши
				$( '#crosshair-label-action-key' ).html( "<div class='action-key-lmb' />" );
			} else {
				$( '#crosshair-label-action-key' ).text( actionKey );
			}
		}
		
		$( '#crosshair-label-text-main' ).html( labelText );
		$( '#crosshair-label-text-progress-overlay' ).html( labelText );
		$( '#crosshair-label-description' ).html( labelDescription );
		
		if ( labelColor != null && labelColor != undefined && labelColor != false ) {
			$( '#crosshair-label-text' ).css( 'color', labelColor );
		} else {
			$( '#crosshair-label-text' ).css( 'color', 'inherit' );
		}
		$( '#crosshair-label' ).show();
	},
	
	// Убрать подсказку возле курсора
	removeLabel : function() {
		$( '#crosshair-label' ).hide();
	},
	
	// Установить прогресс действия (цвет текста слева направо изменится)
	setLabelProgress : function( progress, color ) {
		//console.log( $( '#crosshair-label-text-main' ).width() );
		
		if ( color != false && color != undefined && color != null ) {
			$( '#crosshair-label-text-progress-overlay' ).css( 'color', color );
		} else {
			$( '#crosshair-label-text-progress-overlay' ).css( 'color', '#4CAF50' );
		}
		
		$( '#crosshair-label-text-progress-overlay' ).css( 'width', ( $( '#crosshair-label-text-main' ).width() * progress / 100 ) + 'px' );
		$( '#crosshair-label-text-progress-overlay' ).show();
	},
	
	removeLabelProgress : function() {
		$( '#crosshair-label-text-progress-overlay' ).hide();
	}
};
$( document ).ready( function() {
	Crosshair.init();
} );

File: resources/MTAW/client/crosshair.lua
-- Работает как Cursor, только наоборот
-- Виден всегда, если ни один модуль не отключил его (disable)

--------------------------------------------------------------------------------
--<[ Модуль Crosshair ]>--------------------------------------------------------
--------------------------------------------------------------------------------
Crosshair = {
	isVisible = true;
	disabledBy = {};
	
	init = function()
		addEventHandler( "Main.onClientLoad", root, Crosshair.onClientLoad )
	end;
	
	onClientLoad = function()
		-- Инициализация (создание) прицела
		GUI.sendJS( "Crosshair.enable" )
	
		-- Начальная загрузка формы прицела
		Crosshair.setShape( CFG.gameplay.crosshairShape )
		
		-- Обработка
		addEventHandler( "onClientRender", root, Crosshair.onClientRender )
		
		-- Если изменилась настройка формы прицела, обновляем
		addEventHandler( "Settings.onSettingChanged", resourceRoot, function( categoryName, itemName, oldValue, newValue )
			if ( categoryName == "gameplay" and itemName == "crosshairShape" ) then
				Crosshair.setShape( newValue )
			end
		end );
	end;
	
	-- Отключить отображение прицела
	-- > disabledBy string - причина (но чаще - название модуля), по которой прицел отключается. Пример: "Chat". Это же значение используется в cancelDisabling для отмены отключения прицела
	-- = void
	disable = function( disabledBy )
		if ( Crosshair.disabledBy[ disabledBy ] == nil ) then
			-- Еще не был отменен этим модулем
			Crosshair.disabledBy[ disabledBy ] = true
			
			if ( Crosshair.isVisible ) then
				GUI.sendJS( "Crosshair.setVisible", false )
				Crosshair.isVisible = false;
			end
		end
	end;
	
	-- Отменить отключение прицела
	-- > disabledBy string - причина, по которой прицел был отключен, и которая была использована в disable
	-- = void
	cancelDisabling = function( disabledBy )
		if ( Crosshair.disabledBy[ disabledBy ] ~= nil ) then
			-- Этот модуль ранее отменял прицел, продолжаем
			Crosshair.disabledBy[ disabledBy ] = nil
			
			if ( tableRealSize( Crosshair.disabledBy ) == 0 ) then
				-- Больше никто не запрещает отображение прицела
				GUI.sendJS( "Crosshair.setVisible", true )
				Crosshair.isVisible = true;
			end
		end
	end;
	
	-- Установить подсказку возле прицела
	-- > labelText string - текст подсказки
	-- > labelDescription string - описание (находится под текстом и используется для уточнения)
	-- > labelColor string / nil - цвет подсказки
	-- > showActionKey bool / nil - если true, слева от подсказки будет показана кнопка действия actionKey
	-- > actionKey string / nil - кнопка действия. Например, "E" или "RMB" (правая кнопка мыши). По умолчанию E
	-- = void
	setLabel = function( labelText, labelDescription, labelColor, showActionKey, actionKey )
		if labelColor == nil then labelColor = false end
		
		GUI.sendJS( "Crosshair.setLabel", labelText, labelDescription, labelColor, showActionKey, actionKey )
	end;
	
	-- Спрятать подсказку возле прицела
	-- = void
	removeLabel = function()
		GUI.sendJS( "Crosshair.removeLabel" )
	end;
	
	-- Показать прогресс действия (в процентах). Слева направо цвет текста label будет меняться
	-- > progress number - прогресс действия (от 0 до 100)
	-- > color string / nil - цвет, которым будет заполняться label (по умолчанию зеленый, #4CAF50)
	-- = void
	setLabelProgress = function( progress, color )
		GUI.sendJS( "Crosshair.setLabelProgress", progress, color )
	end;
	
	-- Убрать прогресс действия с подсказки
	-- = void
	removeLabelProgress = function()
		GUI.sendJS( "Crosshair.removeLabelProgress" )
	end;
	
	-- Установить форму прицела
	-- > shape string - форма прицела (по умолчанию: "circle")
	-- = void
	setShape = function( shape )
		--Debug.info( "Установлена форма прицела:", shape )
		-- TODO 
	end;
	
	----------------------------------------------------------------------------
	--<[ Обработчики событий ]>-------------------------------------------------
	----------------------------------------------------------------------------
	
	onClientRender = function()
		-- Не показываем прицел, если игрок целится из оружия
		if ( getControlState( "aim_weapon" ) ) then
			Crosshair.disable( "AimingWeapon" )
		else
			Crosshair.cancelDisabling( "AimingWeapon" )
		end
	end;
}
addEventHandler( "onClientResourceStart", resourceRoot, Crosshair.init )